### DISCUSSION TITLE
Reverse proxy to MQTT broker over Websockets

### DISCUSSION THREAD

**Discussion by tukusejssirs** [nginx-proxy/nginx-proxy#2035] [Q&A]
Related: #1670, #128, #1490

---

I want to reverse proxy to MQTT broker over Websockets using `nginx-proxy` in Docker container, but I can’t make it work.

It works with `nginx` Docker container (`aedes_ws` is another Docker container in which an MQTT broker running over Websockets on port `8000`):

```nginx
location /api/mqtt {
   proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
   proxy_set_header Host $host;
   proxy_pass http://aedes_ws:8000;
   proxy_http_version 1.1;
   proxy_set_header Upgrade $http_upgrade;
   proxy_set_header Connection "upgrade";
}
```

I tried to add either of the following configs to `nginx-proxy`. I tried to put both configs to `/etc/nginx/vhost.d/aedes_ws.conf`, `/etc/nginx/conf.d/aedes_ws.conf` and `/etc/nginx/proxy.conf` (of course, one config to one place at a time).

```nginx
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header Host $host;
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection 'upgrade';
```

```nginx
proxy_http_version 1.1;
proxy_buffering off;
proxy_set_header Host $http_host;
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection $proxy_connection;
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header X-Forwarded-Proto $proxy_x_forwarded_proto;
proxy_set_header X-Forwarded-Ssl $proxy_x_forwarded_ssl;
proxy_set_header X-Forwarded-Port $proxy_x_forwarded_port;
proxy_set_header X-Forwarded-Path $request_uri;

# Mitigate httpoxy attack
proxy_set_header Proxy '';
```

I try to connect to it in Node.js using the following code:

```js
const client = require('mqtt').connect(`ws://127.0.0.1:8000/api/mqtt`)
client.on('connect', () => console.log('MQTT connected'))
client.on('reconnect', () => console.log('MQTT reconnecting'))
client.on('offline', () => console.log('MQTT offline'))
```

However, it fails to connect via `nginx-proxy`, either that MQTT broker is offline (when config is in `/etc/nginx/proxy.conf` or `/etc/nginx/vhost.d/aedes_ws.conf`) with `ECONNREFUSED` error (connection refused; when config is in `/etc/nginx/conf.d/aedes_ws.conf`).

I use this env variables in `docker-compose` for `aedes_ws` container:

```ini
VIRTUAL_DEST: /
VIRTUAL_HOST: localhost
VIRTUAL_PATH: /api/mqtt  # I also tried `/api/mqtt/`
VIRTUAL_PORT: 8000
```

How can I make it work?

Thanks in advance for any help! :pray:

**Comment by gastonmastra** [comment]
Hi!
Did you fix that?
